### PR TITLE
Fix HalModuleParser.parsePrefix() when parsing valid compressed binaries

### PR DESCRIPTION
**Pull request by avtolstoy**
Current version of binary-version-reader may provide invalid module prefix for valid compressed modules due to hardware platform-specific heuristics.

This causes errors in Device Service and disables compression and module combination when performing OTAs:
```
RangeError: Invalid size of the module data\
 at combineModules (/usr/src/app/node_modules/binary-version-reader/lib/moduleEncoding.js:405:10)\
 at runMicrotasks (<anonymous>)\
 at processTicksAndRejections (internal/process/task_queues.js:93:5)\
 at async Flash._combineModuleBinaries (/usr/src/app/lib/behaviors/Flash.js:475:18)\
 at async Flash._preprocessModuleBinary (/usr/src/app/lib/behaviors/Flash.js:394:20)\
 at async Flash.flash (/usr/src/app/lib/behaviors/Flash.js:130:12)"
```

The following binary after compression will not be correctly parsed:
[tracker-system-part1@3.1.0-alpha.2.bin](https://media.clubhouse.io/api/attachments/files/clubhouse-assets/57ec0734-ad50-41a7-86a4-ad1034433bfe/60ad8015-63dd-4e61-b6be-c2df329c062a/tracker-system-part1@3.1.0-alpha.2.bin)

Compressed:
[compressed_tracker-system-part1@3.1.0-alpha.2.bin](https://media.clubhouse.io/api/attachments/files/clubhouse-assets/57ec0734-ad50-41a7-86a4-ad1034433bfe/60ad800b-564b-4426-ba54-08a2b5842ff3/compressed_tracker-system-part1@3.1.0-alpha.2.bin)

```console
 CRC is ok (35df8d60)
 Compiled for unknown platform (ID: 19388)
 This is an unknown module number 182 at version 8993
 It depends on undefined number 20 at version 62570
```

This should instead be:
```console
 CRC is ok (35df8d60)
 Compiled for tracker
 This is a system module number 1 at version 3101
 It depends on a bootloader number 0 at version 1006
 It also depends on a network coprocessor (NCP) module number 0 at version 7
```

### CODE REVIEW

File: lib/HalModuleParser.js [hunk 1]
```diff
@@ -231,81 +231,150 @@ HalModuleParser.prototype = {
...
+	 * Validates whether the module binary has a valid
+	 * module prefix at a certain offset.
 	 *
-	 * This changes based on whether or not we're in the bootloader,
-	 * and we attempt to validate the result based on the assumption
-	 * that firmware addresses tend to start with hex '80' since
-	 * they're stored after
-	 * 0x0800xxxx, etc.
+	 * @param {Object} fileBuffer
+	 * @param {number} offset
```
[the remark this answers is not in the document]
Comment (wraithan) [reply lib/HalModuleParser.js:238]: Also, when I search for uses of `_validateModulePrefix` I don't actually see this parameter used. So it seems like we're always in non-strict mode?

**Review by wraithan — CHANGES_REQUESTED** [review]
Overall the code looks good, reads a lot better and I appreciate the code clean up you've done.

I'm a bit 🤔 about the `strict` option to `_validateModulePrefix` and would like to discuss before approval.

File: lib/HalModuleParser.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:238]: Yeah, the argument not used anywhere, so perhaps we can default to the non-strict behavior and remove it for now entirely. The main problem is that there is no sure-proof way of saying whether what we've found is an actual module prefix or something that just looks like it. This also becomes more problematic if we take into account modules that are not explicitly supported currently in binary-version-reader (e.g. imagine a case where we introduce new platform id or module type, this would cause those checks to fail lowering the score far more than just below max - 1). See a test on `unknown.bin` for an example of that.

There are also other corner cases which I did not want to break as of this PR that may cause some of the checks to fail. For example `RC4_bootloader_pad_BM-09.bin` used in tests will not pass start address - end address == binary size, as the binary is padded.

[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:238]: We can also get rid of the score altogether and just output the result of each particular test case that does not immediately fail the validation as a bool in an object and just have the caller look into that?

[the remark this answers is not in the document]
Comment (wraithan) [reply lib/HalModuleParser.js:238]: I'm actually really comfortable with the score aspect and get that we need to heuristic it, it's more the boolean pass/fail based on less than 100% that's weird to me. Totally happy with us having a score and choosing best scoring address.

[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:238]: Ah, okay, I see what you mean. Yeah, agreed, looks weird.

Let's get rid of it and suffer a hit of having to walk over the full 8KB in case of malformatted or not fully supported binaries (e.g. unknown platform id etc) :+1: ?

[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:238]: Fixed.

File: lib/HalModuleParser.js [hunk 2]
```diff
@@ -231,81 +231,149 @@ HalModuleParser.prototype = {
...
-		// of the bootloader....  we'll need some extra sanity checks somehow later if hardware changes dramatically.
-		var mightBeUser = (userModuleStartAddy.indexOf("80") === 0) || (userModuleStartAddy === 'd4000');
-		var mightBeSystem = (sysModuleStartAddy.indexOf("80") === 0);
-		var isCore = ((userModuleStartAddy.indexOf("20") === 0) && (coreModuleStartAddy === "8005000"));
-		var mightBeBluz = (bluzModulesPlatformID == 103 && bluzModuleStartAddy < 262144);
-		var isGen3 = (meshModule.platformID >= 12 && meshModule.platformID <= 30) && (meshModule.moduleStartAddy === '30000' || meshModule.moduleStartAddy === 'f4000') &&
-			(userModuleStartAddy.indexOf("20") == 0);   // stack is located in ram at 0x2000000
-
-		if (isCore) {
-			return 0x10C;
```
Comment (sergeuz) [inline lib/HalModuleParser.js:296]: Oh yes. Thank you so much for doing this.

File: lib/HalModuleParser.js [hunk 3]
```diff
@@ -231,81 +231,150 @@ HalModuleParser.prototype = {
...
+		// This is simply an optimization to try a few preferred locations first
+		// 0x184 is the size of Gen 2 MCU (STM32F2) interrupt vector table
+		// 0x200 is the size of Gen 3 MCU (nRF52840) interrupt vector table
+		// FIXME: Some of the deprecated platforms should be removed
+		const preferredOffsets = [0,
+			0x200, // Gen 3
+			0x184, // Gen 2
+			0xC0, // Bluez
+			0x10C, // Core
+		];
```
[the remark this answers is not in the document]
Comment (sergeuz) [reply lib/HalModuleParser.js:342]: At first I was going to suggest adding a zero offset to the list of preferred ones, but then I noticed that it's actually there, it just didn't deserve its own line in the array definition. I'd add a comment that the module header typically starts at 0 offset if it's a user part or other non-system/bootloader module.

File: lib/HalModuleParser.js [hunk 4]
```diff
@@ -23,9 +23,9 @@ var when = require('when');
 var pipeline = require('when/pipeline');
 var utilities = require('./utilities.js');
 var buffers = require('h5.buffers');
```
Comment (sergeuz) [inline lib/HalModuleParser.js:25]: This module is no longer used in this file. I'd also check if we even need this dependency at all, given the changes in this PR.

**Review by sergeuz — APPROVED** [review]
What an awesome PR :+1:

File: lib/HalModuleParser.js (hunk 4 above)
Comment (avtolstoy) [reply lib/HalModuleParser.js]: You are right, no longer required. Removed the dependency.

File: lib/HalModuleParser.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:347]: Moved 0 to a separate line and added an additional comment.

File: lib/HalModuleParser.js [hunk 5]
```diff
@@ -231,81 +231,149 @@ HalModuleParser.prototype = {
...
+		if (typeof endAddr === 'string') {
+			endAddr = Number.parseInt(endAddr, 16);
+		}
+		if (endAddr <= startAddr) {
+			return result;
+		}
+		const expectedSize = endAddr - startAddr + 4 /* CRC-32 */;
+		if (fileBuffer.length < expectedSize) {
+			return result;
+		}
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:267]: https://github.com/particle-iot/binary-version-reader/pull/41/files#diff-1191fb254480711d2964abf2c8ce9727c95f0e24d7a032142ff18c792b49b14bR306

File: lib/HalModuleParser.js [hunk 6]
```diff
@@ -231,81 +231,149 @@ HalModuleParser.prototype = {
...
+		// Module size according to the module prefix exactly matches (should pass for non-padded binaries)
+		if (fileBuffer.length === expectedSize) {
+			result.score++;
+		}
+		// Module flags are known
+		const allFlags = Object.values(ModuleInfo.Flags).reduce((a, b) => (a | b), 0);
+		if ((prefixInfo.moduleFlags & (~allFlags)) === 0) {
+			result.score++;
+		}
```
[the remark this answers is not in the document]
Comment (avtolstoy) [reply lib/HalModuleParser.js:315]: > Also, the fact that start/end addresses are in the internal flash (where applicable) that was checked in the old code was a good indication that we're dealing with a valid module, IMO.

I think we should be able to get away without these kinds of platform-specific checks and probably should if we can. It's also impossible to implement such checks properly for Gen 3 platforms as addresses in the internal flash are generic 0 to 1MB.

> Non-zero version numbers (of the module itself and its dependencies) can be another 0.25 score points :slightly_smiling_face:

As discussed in Slack, let's keep things as-is. Given all the other checks it's anyway highly likely that the versions will not be zero.

**Review by wraithan — APPROVED** [review]
Thanks for the updates, looks great
